content/blog/scheduling-in-react/index.md

```diff
@@ -0,0 +1,243 @@
+
+Let’s take another look at the Performance tab together:
+
+![Screenshot of Chrome DevTools that shows that React breaks the rendering work down into small chunks. All frames can be drawn very quickly, although analytics still happen in the middle.](devtools-normal.png)
+
+We see that the long-running tasks are now broken down into smaller ones that can be completed within a single frame. The red triangles that indicate frame drops are also gone.
+
+However, one thing that is still not ideal is that the analytics notification (highlighted in the above screenshot) is still executed with the rendering work. Since the users of our app do not see this task, we can assign it an even lower priority:
+
+```js{3-5,18-22}
+import {
+  unstable_next,
+  unstable_LowPriority,
+  unstable_runWithPriority,
+  unstable_scheduleCallback
+} from "scheduler";
+
+function SearchBox(props) {
+  const [inputValue, setInputValue] = React.useState();
+
+  function handleChange(event) {
+    const value = event.target.value;
+
+    setInputValue(value);
+    unstable_next(function() {
+      props.onChange(value);
+    });
+    unstable_runWithPriority(unstable_LowPriority, function() {
+      unstable_scheduleCallback(function() {
+        sendAnalyticsNotification(value);
+      });
+    });
+  }
+
+  return <input type="text" value={inputValue} onChange={handleChange} />;
+}
+```
+
+If we take another look at the Performance tab with this change and zoom toward the end, we’ll see that our analytics are now sent after all rendering work has completed, and so the total work in our app is perfectly scheduled:
+
+![Screenshot of Chrome DevTools that show that React breaks the rendering work down into small chunks. Analytics are sent at the end after all rendering work has completed.](devtools-normal-and-low.png)
+
+Try it out:
+
+<iframe src="https://codesandbox.io/embed/v0lxm0xlzl?fontsize=14&hidenavigation=1&view=preview" style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;" sandbox="allow-modals allow-forms allow-popups allow-scripts allow-same-origin"></iframe>
+
+## Limitations of the Scheduler
+
+With the Scheduler, it’s possible to control when certain callbacks can be executed. It is built deep into the latest React implementation and works out of the box with Concurrent mode.
+
+That said, tere are two limitations of the Scheduler:
+
+1. **Resource Fighting.** The Scheduler tries to use all of the resources available. This causes issues if multiple instances of a scheduler run on the same thread and compete for resources. We need to ensure that all parts of our application will use the same instance.
+2. **Interleaving tasks with browser work.** Since the Scheduler runs in the browser, it can only do what the browser allows us to. Document lifecycles like rendering or garbage collection can interfere with the work in an uncontrollable way.
```

Comment:
I don't love the use of the word "interleaving" here. If it's common in the community it can stay but maybe there's a better word?

"**Interleaving tasks with browser work.**"

Reply:
I think I copied this word from some of the resources for this post but I don't feel it's super clear either. What I try to say is that the browser does some work on its own (the document lifecycles) and it also needs to run the callbacks we scheduled. So a sophisticated solution would take both types of work into account and balance them accordingly.

Maybe "Balance user defined tasks with browser work"?

Reply:
I would expect the scheduler to keep track of its time budget, and postpone calling callbacks if this budget is reduced due to the browser doing something else (or due to any unknown reason). But apparently we still have to manage this ourselves when using the scheduler. Have I understood this correctly?

Reply:
@miguelcalderon The thing is that the Scheduler doesn't know when the browser is doing something else. It can only measure the things that the browser gives us which is the rAF callback before a frame is rendered which can be used to find out how big the frame budget is. 

So if the browser is doing garbage collection at a random point in time, the Scheduler's timer will be off.

Do you think this needs further explanation? I plan on adding one paragraph that explains how the Scheduler works under the hood without going into too much detail, something like this:

> Under the hood, the Scheduler uses platform APIs like `requestAnimationFrame` to infer the frame budget available. It will then schedule callbacks that run _after_ a frame is drawn and will time the callbacks so that they can return before the next frame is rendered.

Looking at this sentence though it seems like it's too much information cramped into a one paragraph and I'm not sure if it's important to know how the Scheduler works under the hood if you just want to learn about the problem space and how it can be solved. 

Maybe I can add this to the conclusion section:

> If you're curious how the Scheduler works under the hood, take a look at the [current implementation](). I plan to go into more detail once the Scheduler is considered stable.